server/src/main/java/org/elasticsearch/index/mapper/JsonFieldParser.java

```diff
@@ -0,0 +1,64 @@
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentParserUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A helper class for {@link JsonFieldMapper} parses a JSON object
+ * and produces an indexable field for each leaf value.
+ */
+public class JsonFieldParser {
+    private final MappedFieldType fieldType;
+    private final int ignoreAbove;
+
+    JsonFieldParser(MappedFieldType fieldType,
+                    int ignoreAbove) {
+        this.fieldType = fieldType;
+        this.ignoreAbove = ignoreAbove;
+    }
+
+    public List<IndexableField> parse(XContentParser parser) throws IOException {
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT,
+            parser.currentToken(),
+            parser::getTokenLocation);
+
+        List<IndexableField> fields = new ArrayList<>();
+        int openObjects = 1;
+
+        while (true) {
+            if (openObjects == 0) {
+                return fields;
+            }
+
+            XContentParser.Token token = parser.nextToken();
+            assert token != null;
+
+            if (token == XContentParser.Token.START_OBJECT) {
+                openObjects++;
+            } else if (token == XContentParser.Token.END_OBJECT) {
+                openObjects--;
+            } else if (token.isValue()) {
+                String value = parser.text();
+                addField(value, fields);
+            } else if (token == XContentParser.Token.VALUE_NULL) {
+                String value = fieldType.nullValueAsString();
+                if (value != null) {
+                    addField(value, fields);
+                }
+            }
+        }
+    }
+
+    private void addField(String value, List<IndexableField> fields) {
+        if (value.length() <= ignoreAbove) {
+            fields.add(new Field(fieldType.name(), new BytesRef(value), fieldType));
```

Comment:
For now I opted not to index a special stored field with the whole JSON blob (as we discussed in https://github.com/elastic/elasticsearch/pull/33795#issuecomment-422535360), but it's something I'll keep in mind.